### ISSUE TITLE
We need a way to filter the metrics that match a prefix, instead of filtering all the ones that don't match a prefix

### ISSUE THREAD

**Issue by replay** [grafana/carbon-relay-ng#363]
There's a customer who wants to only send metrics to a route which don't match a given regex. So they'd need the inverted functionality of the current `prefix` parameter. 
I think the best solution would be to add something like a bool called `invert_filter`, which inverts how the filtering for `prefix`, `substr` and `regex` works.

**Comment by DanCech** [comment]
`invert_filter` seems like a non-intuitive way to achieve this configuration.

One alternative that comes to mind would be to support separate `not_substr`, `not_prefix` and `not_regex` settings, which would have the benefit of being able to be combined with the normal filters like `prefix = foo.` `not_substr = bar`.

Another option would be to support a modifier in the value itself, but that might also be non-intuitive.

**Comment by Dieterbe** [comment]
not so much non-intuitive, but rather non-practical. one benefit of the matchers is that you can do regex and substring and prefix. having an invert setting apply to all of them seems needlessly constraining.

we can find some inspiration in prometheus and graphite tag selectors, e.g. `=~`, `!~` for regex though it's not a good match. afaik these selectors either do regex or full equality (which for metrics routing doesn't make much sense), and there's no good syntax for prefix and substring (though we could specify a prefix as `^prefix' and a substring as '.*substring.*' but that would be tricky to parse out properly.

Also, this customer you're talking about, weren't they really looking for a "first route match" behavior (across routes)? Your suggestion makes sense, and may be the easiest way to achieve it, but let's make sure we don't loose sight of the real goal.

**Comment by woodsaj** [comment]
I think this type of functionality would be better handled with the existing 'route' and 'destination' concepts.  However this would require major work in crng to implement these concepts correctly, eg make GrafanaNet, Kafka, Pub/Sub etc... function as destinations and not as routes like they currently do.

By then adding a simple "null" destination type you could just use a 'sendFirstMatch' route to support the original goals of this issue ` only send metrics to a route which don't match a given regex`.   eg.

```
## discard metrics  
[[destination]]
key=nullRoute
type=Null

# send to GrafanaCloud
[[destination]]
key = 'grafanaCloud'
type = 'grafanaNet'
addr = https://grahite-us-central1.grafana.net

[[route]]
# send everything that doesnt match foo.*.baz.* to GrafanaCloud
key = 'fooBaz'
type = 'sendFirstMatch'
destinations = [
  'nullRoute regex=^foo\..*\.baz\..*',
  'grafanaCloud'
]
```

This example config is purely hypothetical.  It allows `destinations` to be defined and changes the route.destinations to use a destination name followed by 'prefix', 'substr' and 'regex' options.   While this works for this specific example, more thought would be needed for an overhaul of the configuration schema.

in-lieu of a total overhaul of routes/destinations, i prefer @DanCech's approach of not_substr, not_prefix and not_regex as this gives the greatest amount of flexibility.    The only downside for that is it has to be implemented for all routes and destinations.


finally, if we just want a quick "hack" to support inverse match for regexes, then i would just use a custom regex flag (passed using `(?flags)` ) that we look for and strip out from the regex before processing.  eg if regex starts with something like `(?!)` then match everything doesnt match the regex.

**Comment by replay** [comment]
>  i would just use a custom regex flag (passed using (?flags) ) that we look for and strip out from the regex before processing

That's an interesting idea, I haven't thought of that. As a quick fix, I think that makes sense

**Comment by replay** [comment]
> Also, this customer you're talking about, weren't they really looking for a "first route match" behavior (across routes)?

Yes, that's what the customer was asking for. Although, after chatting, I had the impression the customer is only trying to achieve the goal of being able to send everything except certain prefixes into a route. So I think any solution that allows them to do this is satisfactory.

**Comment by Dieterbe** [comment]
i believe this was implemented in #422 and #423
